spring-boot/proxy/src/main/java/eu/coatrack/proxy/security/LocalApiKeyManager.java

```diff
@@ -0,0 +1,125 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+import eu.coatrack.api.ApiKey;
+import eu.coatrack.proxy.security.exceptions.LocalApiKeyListWasNotInitializedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.scheduling.annotation.EnableAsync;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+import org.springframework.util.Assert;
+
+import javax.annotation.PostConstruct;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Provides a local cache for API keys, allowing the gateway to validate API keys without connection to CoatRack
+ * admin application. Caches (and periodically refreshes) a list of all API keys assigned to all services offered
+ * by the gateway.
+ *
+ * @author Christoph Baier
+ */
+
+@EnableAsync
+@Service
+public class LocalApiKeyManager {
+
+    private static final Logger log = LoggerFactory.getLogger(LocalApiKeyManager.class);
+
+    private List<ApiKey> localApiKeyList = new ArrayList<>();
+    private LocalDateTime deadlineWhenOfflineModeShallStopWorking = LocalDateTime.MIN;
+
+    private final ApiKeyFetcher apiKeyFetcher;
+    private final long numberOfMinutesTheGatewayShallWorkInOfflineMode;
+
+    private GatewayMode lastModeDisplayedInLog = GatewayMode.OFFLINE;
+
+    public LocalApiKeyManager(
+            ApiKeyFetcher apiKeyFetcher,
+            @Value("${number-of-minutes-the-gateway-shall-work-in-offline-mode}") long minutesInOfflineMode) {
+        this.apiKeyFetcher = apiKeyFetcher;
+        this.numberOfMinutesTheGatewayShallWorkInOfflineMode = minutesInOfflineMode;
+    }
+
+    public ApiKey getApiKeyEntityByApiKeyValue(String apiKeyValue) {
+        log.debug("Trying to get the API key with the value {} from the local list.", apiKeyValue);
+        updateGatewayMode(GatewayMode.OFFLINE);
+
+        return localApiKeyList.stream().filter(
+                apiKeyFromLocalList -> apiKeyFromLocalList.getKeyValue().equals(apiKeyValue)
+        ).findFirst().orElse(null);
+    }
+
+    public boolean isOfflineWorkingTimeExceeded() {
+        if (deadlineWhenOfflineModeShallStopWorking == LocalDateTime.MIN)
+            throw new LocalApiKeyListWasNotInitializedException("The offline mode does not work without a local " +
+                    "copy of the API key list.");
+        return LocalDateTime.now().isAfter(deadlineWhenOfflineModeShallStopWorking);
+    }
+
+    @Async
+    @PostConstruct
+    @Scheduled(fixedRateString = "${local-api-key-list-update-interval-in-millis}")
+    public void updateLocalApiKeyList() {
```

Comment:
Maybe use a more clear name like:
`refreshLocalApiKeyCacheWithApiKeysFromAdmin`
or
`getApiKeysFromAdminAndUpdateLocalCache`

Reply:
I preferred the first option. Done.